src/stream/src/executor/backfill/no_shuffle_backfill.rs

```diff
@@ -163,16 +163,20 @@ where
                 // It is finished, so just assign a value to avoid accessing storage table again.
                 false
             } else {
-                let snapshot = Self::snapshot_read(
-                    &self.upstream_table,
-                    init_epoch,
-                    None,
-                    false,
-                    self.chunk_size,
-                    &mut builder,
-                );
-                pin_mut!(snapshot);
-                snapshot.try_next().await?.unwrap().is_none()
+                let snapshot_is_empty = {
+                    let snapshot = Self::snapshot_read(
+                        &self.upstream_table,
+                        init_epoch,
+                        None,
+                        false,
+                        &mut builder,
+                    );
+                    pin_mut!(snapshot);
+                    snapshot.try_next().await?.unwrap().is_none()
+                };
+                let snapshot_buffer_is_empty = builder.is_empty();
+                builder.clear();
+                snapshot_is_empty && snapshot_buffer_is_empty
```

Comment:
Here's the code of interest cc @chenzl25 .

To reproduce it:
1. Initial snapshot read occurs, but does not produce a chunk. Instead it only fills up the builder's buffer. Let's say row A is now in this buffer.
2. When backfill occurs later, snapshot is refreshed, it will read row A again.
3. Now builder's buffer has 2x row A, if it was not cleared after 1.

The test was originally found in arrangement backfill.
I think the same issue could be encountered in `no_shuffle_backfill` as well, but don't have a test case for it.

Reply:
However, I can't construct such a case manually. Actually, I want to evaluate the scope of impact.

Reply:
Since ` snapshot.try_next().await?.unwrap().is_none()` will wait until a chunk returns. The builder's buffer should contain no extra row I think.

Reply:
OH it must be arrangement backfill specific.
You raise a good point. I think it only happens in arrangement backfill, and not here, because snapshot_read per vnode returns a stream of chunks.

To be precise, the following scenario occurs:
1. Arrangement backfill does initial snapshot read, this occurs for multiple vnode streams.
2. Perhaps 1 vnode stream returns a chunk, so it continues. Concurrently, another vnode stream is in process of constructing its chunk, but did not finish it.
3. Then what happens is that for the first vnode stream which returned a chunk, it will clean up the corresponding builder.
4. But for the subsequent vnode stream which did not complete it, it will still have rows in the buffer, which need to be cleaned.

This does not apply here, with a single snapshot read stream.

I will remove the "bugfix" in this PR, and just address bugen's comments on API changes here.

Reply:
Oh but actually could there be an (unlikely) scenario where `iter_chunks` continues executing on row_iter?

i.e. the following sequence of events:
```rs
                    let snapshot = Self::snapshot_read(
                        &self.upstream_table,
                        init_epoch,
                        None,
                        false,
                        &mut builder,
                    );
                    pin_mut!(snapshot);
                    snapshot.try_next().await?.unwrap().is_none() -> first chunk is processed.
  --> here the next snapshot chunk is being processed asynchronously.
} --> drop is called before the chunk is finished processing
```

Reply:
If we don't call the `next`, I think it is safe. Anyway, clear the buffer at the end LGTM.